File: packages/email/src/canspam.ts
export type CanSpamConfig = {
  postalAddress: string;
  unsubscribeUrl?: string;
};

export function assertCanSpamCompliance(config: CanSpamConfig): void {
  if (!config.postalAddress || config.postalAddress.trim() === "") {
    throw new Error(
      "CAN-SPAM compliance requires a physical postal address. Set postalAddress in EmailClientConfig.",
    );
  }
  const lower = config.postalAddress.toLowerCase();
  if (
    (lower.includes("[") && lower.includes("]")) ||
    lower.includes("placeholder") ||
    lower.includes("todo")
  ) {
    throw new Error(
      `CAN-SPAM postalAddress looks like a placeholder: "${config.postalAddress}". Set a real address.`,
    );
  }
}

export function buildListUnsubscribeHeaders(unsubscribeUrl: string): Record<string, string> {
  return {
    "List-Unsubscribe": `<${unsubscribeUrl}>`,
    "List-Unsubscribe-Post": "List-Unsubscribe=One-Click",
  };
}

File: py/ventora_email/src/ventora_email/canspam.py
from __future__ import annotations


def assert_can_spam_compliance(postal_address: str) -> None:
    """Raise ValueError if postal_address is missing or looks like a placeholder."""
    if not postal_address or not postal_address.strip():
        raise ValueError("CAN-SPAM requires a valid postal address")
    lower = postal_address.lower()
    bracketed = "[" in postal_address and "]" in postal_address
    if bracketed or "placeholder" in lower or "todo" in lower:
        raise ValueError(f"CAN-SPAM postal address appears to be a placeholder: {postal_address!r}")


def build_list_unsubscribe_headers(unsubscribe_url: str) -> dict[str, str]:
    """Build RFC 8058 List-Unsubscribe headers for one-click unsubscribe."""
    return {
        "List-Unsubscribe": f"<{unsubscribe_url}>",
        "List-Unsubscribe-Post": "List-Unsubscribe=One-Click",
    }
